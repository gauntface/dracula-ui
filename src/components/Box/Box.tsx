import cx from 'classnames/dedupe'
import React, { AllHTMLAttributes } from 'react'
import { colors } from '../../base/colors'
import { spacingClasses, SpacingPropType } from '../../base/spacing'

type Element = HTMLElementTagNameMap

export const roundedBorders = {
  none: 'drac-rounded-none',
  sm: 'drac-rounded-sm',
  base: 'drac-rounded',
  lg: 'drac-rounded-lg',
  xl: 'drac-rounded-xl',
  '2xl': 'drac-rounded-2xl',
  '3xl': 'drac-rounded-3xl',
  full: 'drac-rounded-full',
}

/**
 * Box Props
 */
export interface BoxProps<K extends keyof Element = 'div'>
  extends AllHTMLAttributes<K> {
  /** The background color. */
  color?: keyof typeof colors

  /** The border radius. */
  rounded?: keyof typeof roundedBorders

  /** Dracula UI standard spacing properties. */
  spacing?: SpacingPropType

  /** The HTML element to be used */
  as?: K
}

/**
 * Box is the most primitive component of Dracula UI.
 * Using Box allows for consumers of the components library to compose
 * more complex patterns, components, and UIs.
 *
 * Box includes built-in Color and Spacing properties that make building
 * complex components convenient and consistent.
 */
export function Box<T extends keyof Element>(props: BoxProps<T>) {
  const finalProps = {
    ...props,
    className: cx(
      `drac-box`,
      props.color && colors[props.color],
      props.rounded && roundedBorders[props.rounded],
      ...spacingClasses(props.spacing),
      props.className
    )
  }

  const as = finalProps.as ?? 'div'
  return React.createElement(as, finalProps, props.children)
}

Box.displayName = 'Box'
