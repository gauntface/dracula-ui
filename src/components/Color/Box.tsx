import React, { HTMLAttributes } from 'react'

export const colors = {
  black: 'drac-bg-black',
  white: 'drac-bg-white',
  cyan: 'drac-bg-cyan',
  green: 'drac-bg-green',
  orange: 'drac-bg-orange',
  pink: 'drac-bg-pink',
  purple: 'drac-bg-purple',
  red: 'drac-bg-red',
  yellow: 'drac-bg-yellow',
  purpleCyan: 'drac-bg-purple-cyan',
  yellowPink: 'drac-bg-yellow-pink',
  cyanGreen: 'drac-bg-cyan-green',
  pinkPurple: 'drac-bg-pink-purple',
  animated: 'drac-bg-animated'
}

export interface BoxProps extends HTMLAttributes<HTMLDivElement> {
  color?: keyof typeof colors
}

export const Box: React.FC<BoxProps> = props => {
  const finalProps = {
    ...props,
    className: `drac-box ${colors[props.color ?? 'black']}`
  }

  return <div {...finalProps} />
}
