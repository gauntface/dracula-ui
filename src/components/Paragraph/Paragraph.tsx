import cx from 'classnames/dedupe'
import React from 'react'
import { marginMixin, paddingMixin } from '../../base/spacing'
import { textColors, TextProps, textSizes, textWeights } from '../Text/Text'

/**
 * Paragraph is a semantic component used for blocks of text with
 * semantic meaning.
 *
 * Paragraph accepts all the same customization options as Text.
 */
export const Paragraph: React.FC<TextProps> = (props: TextProps) => {
  const finalProps = {
    ...props,

    className: cx(
      `drac-text`,
      textSizes[props.size ?? 'medium'],
      textWeights[props.weight ?? 'normal'],
      textColors[props.color ?? 'white'],
      ...paddingMixin(props),
      ...marginMixin(props)
    )
  }

  return <p {...finalProps}>{props.children}</p>
}
