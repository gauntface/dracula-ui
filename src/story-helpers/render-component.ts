import testRenderer from 'react-test-renderer'

export type ComponentExample = {
  code: ReturnType<testRenderer.ReactTestRenderer['toJSON']>
  docs: string
  title: string
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {
  const rendered = testRenderer.create(comp)
  const code = rendered.toJSON()
  return {
    title,
    code,
    docs
  }
}
