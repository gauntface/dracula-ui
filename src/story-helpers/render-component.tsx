import React from 'react'
import Enzyme, { mount,  shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import prettier from 'prettier'
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() })

export type ComponentExample = {
  html: string
  docs: string
  title: string
  react?: any
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {

  
  const html = mount(comp).html()
  const react = toJson(shallow(<div>{comp}</div>))

  const prettyHTML = prettier.format(html, {
    parser: 'html'
  })

  return {
    title,
    html: prettyHTML,
    react,
    docs
  }
}
