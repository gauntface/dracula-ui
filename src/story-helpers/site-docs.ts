import {
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'

// import * as serializer from 'jest-snapshot-serializer-raw'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  describe(`Site: ${componentClass.displayName ?? componentClass.name}`, () => {
    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp(),
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
    })

    const variations = documentation.variations()

    variations.forEach(variation => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp(),
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()
      })
    })
  })
}
