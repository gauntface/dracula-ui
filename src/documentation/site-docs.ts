import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '../story-helpers/render-component'
import React from 'react'
import fs from 'fs-extra'
import { pretty } from '../story-helpers/pretty'
import { getDocGen } from '../documentation/docgen/doc-generator'
import { toDSP } from '../documentation/dsp/component-generator'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(async () => {
      const docGen = getDocGen(name)
      const dsp = toDSP(name, examples, docGen)
      const dspPath = `${process.cwd()}/dsp/data/components/${name}.json`
      await fs.writeFile(dspPath, pretty(JSON.stringify(dsp), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
