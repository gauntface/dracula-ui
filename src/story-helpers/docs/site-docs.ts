import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'
import fs from 'fs'
import { pretty } from '@/story-helpers/pretty'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(() => {
      const path = `${process.cwd()}/examples/${name}.json`
      fs.writeFileSync(path, pretty(JSON.stringify(examples), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
