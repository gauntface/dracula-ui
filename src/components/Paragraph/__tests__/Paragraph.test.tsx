import React from 'react'
import { Paragraph } from '@/components/Paragraph/Paragraph'
import {
  ParagraphBasic,
  ParagraphColors,
  ParagraphSizes,
  ParagraphWeights
} from '@/components/Paragraph/__stories__/Paragraph.stories'
import { docs } from '@/documentation/site-docs'
import { snapshot } from '../../../documentation/render-component'
import { Box } from '../../Box/Box'

docs(Paragraph, {
  basic() {
    return snapshot(
      'Usage',
      () => <Paragraph>basic</Paragraph>,
      'Show a paragraph within an interface.'
    )
  },
  variations() {
    return [
      snapshot(
        'Colors',
        () => (
          <Box>
            <Paragraph color="cyan">cyan</Paragraph>
            <Paragraph color="green">green</Paragraph>
          </Box>
        ),
        'Paragraphs can be customized to use any of the Dracula UI theme colors.'
      ),
      snapshot(
        'Weights',
        () => (
          <Box>
            <Paragraph weight="normal">normal</Paragraph>
            <Paragraph weight="semibold">semibold</Paragraph>
            <Paragraph weight="bold">bold</Paragraph>
          </Box>
        ),
        'Paragraphs can be defined in different font weights.'
      ),
      snapshot(
        'Sizes',
        () => (
          <Box>
            <Paragraph size="xsmall">xsmall</Paragraph>
            <Paragraph size="small">small</Paragraph>
            <Paragraph size="medium">medium</Paragraph>
            <Paragraph size="large">large</Paragraph>
          </Box>
        ),
        'Paragraphs can be customize to use several different sizes.'
      )
    ]
  }
})
