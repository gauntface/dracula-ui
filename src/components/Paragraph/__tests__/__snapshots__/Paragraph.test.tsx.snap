// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site: Paragraph Basic Usage 1`] = `
Object {
  "docs": "Show a paragraph within an interface.",
  "html": "<p class=\\"drac-text drac-text-white\\">basic</p>
",
  "react": "<Paragraph>basic</Paragraph>;
",
  "title": "Usage",
}
`;

exports[`Site: Paragraph Colors 1`] = `
Object {
  "docs": "Paragraphs can be customized to use any of the Dracula UI theme colors.",
  "html": "<div class=\\"drac-box\\">
  <p color=\\"cyan\\" class=\\"drac-text drac-text-cyan\\">cyan</p>
  <p color=\\"green\\" class=\\"drac-text drac-text-green\\">green</p>
</div>
",
  "react": "<Box>
  <Paragraph color=\\"cyan\\">cyan</Paragraph>
  <Paragraph color=\\"green\\">green</Paragraph>
</Box>;
",
  "title": "Colors",
}
`;

exports[`Site: Paragraph Sizes 1`] = `
Object {
  "docs": "Paragraphs can be customize to use several different sizes.",
  "html": "<div class=\\"drac-box\\">
  <p class=\\"drac-text drac-text-xs drac-text-white\\">xsmall</p>
  <p class=\\"drac-text drac-text-sm drac-text-white\\">small</p>
  <p class=\\"drac-text drac-text-white\\">medium</p>
  <p class=\\"drac-text drac-text-lg drac-text-white\\">large</p>
</div>
",
  "react": "<Box>
  <Paragraph size=\\"xsmall\\">xsmall</Paragraph>
  <Paragraph size=\\"small\\">small</Paragraph>
  <Paragraph size=\\"medium\\">medium</Paragraph>
  <Paragraph size=\\"large\\">large</Paragraph>
</Box>;
",
  "title": "Sizes",
}
`;

exports[`Site: Paragraph Weights 1`] = `
Object {
  "docs": "Paragraphs can be defined in different font weights.",
  "html": "<div class=\\"drac-box\\">
  <p weight=\\"normal\\" class=\\"drac-text drac-text-white\\">normal</p>
  <p weight=\\"semibold\\" class=\\"drac-text drac-text-semibold drac-text-white\\">
    semibold
  </p>
  <p weight=\\"bold\\" class=\\"drac-text drac-text-bold drac-text-white\\">bold</p>
</div>
",
  "react": "<Box>
  <Paragraph weight=\\"normal\\">normal</Paragraph>
  <Paragraph weight=\\"semibold\\">semibold</Paragraph>
  <Paragraph weight=\\"bold\\">bold</Paragraph>
</Box>;
",
  "title": "Weights",
}
`;
